File: sdk/python/src/conduyt/modules/dht.py
"""CONDUYT DHT Module"""

from __future__ import annotations

import struct
from dataclasses import dataclass
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from conduyt.device import ConduytDevice

from conduyt.core.constants import CMD


@dataclass
class DHTReading:
    temperature: float
    humidity: float


class ConduytDHT:
    def __init__(self, device: ConduytDevice, module_id: int = 0) -> None:
        self._device = device
        self._mod_id = module_id

    async def begin(self, pin: int, sensor_type: int = 22) -> None:
        payload = bytes([self._mod_id, 0x01, pin, sensor_type])
        await self._device._send_command(CMD.MOD_CMD, payload)

    async def read(self) -> DHTReading:
        payload = bytes([self._mod_id, 0x02])
        resp = await self._device._send_command(CMD.MOD_CMD, payload)
        if len(resp) >= 9:  # module_id(1) + 2 floats(8)
            temp, hum = struct.unpack_from("<ff", resp, 1)
            return DHTReading(temperature=temp, humidity=hum)
        return DHTReading(temperature=0.0, humidity=0.0)

File: sdk/swift/Sources/ConduytKit/Modules/DHT.swift
import Foundation

/// One DHT temperature/humidity sample.
public struct DHTReading: Equatable {
    public let temperatureC: Float
    public let humidityPct: Float
}

/// Typed wrapper for the CONDUYT DHT11/DHT22 module.
@available(iOS 15.0, macOS 12.0, *)
public class ConduytDHT {
    private let device: ConduytDevice
    private let moduleID: UInt8

    public init(device: ConduytDevice, moduleID: UInt8) {
        self.device = device
        self.moduleID = moduleID
    }

    /// Initialize the sensor. kind=11 for DHT11, 22 for DHT22.
    public func begin(pin: UInt8, kind: UInt8 = 22) async throws {
        _ = try await device.modCmd(Data([moduleID, 0x01, pin, kind]))
    }

    /// Trigger a sample. Firmware echoes module id, then two LE float32s.
    public func read() async throws -> DHTReading {
        let resp = try await device.modCmd(Data([moduleID, 0x02]))
        guard resp.count >= 9 else {
            throw ConduytDeviceError.nak(code: 0xFE, name: "SHORT_RESPONSE")
        }
        let temp = resp.readLE(float32At: 1)
        let hum  = resp.readLE(float32At: 5)
        return DHTReading(temperatureC: temp, humidityPct: hum)
    }
}
